### DISCUSSION TITLE
How do I create cards/callouts of the same height and positioned next to each other?

### DISCUSSION THREAD

**Discussion by joelostblom** [quarto-dev/quarto-cli#9213] [Q&A] [answered]
### Description

In Jupyter Book, I can start a page with an overview of important information laid out next to each other horizontally using [a grid of cards](https://jupyterbook.org/en/stable/content/components.html#create-grids-of-cards):

![image](https://github.com/quarto-dev/quarto-cli/assets/4560057/bd1454aa-2758-494b-93ac-5ef9b8ae46bd)

I can't seem to find something similar in Quarto. The closest I have gotten is using `layout-col=2`, together with either callouts or cards:

![image](https://github.com/quarto-dev/quarto-cli/assets/4560057/b90d21ae-16f5-4750-ac4b-865b7550d26e)

As you can see in the image, there are a few significant issues:

1. Most importantly, each card/callout is of different height, which ruins the impression that they are part of the same component. Can I control the height to expand to the available height within the column layout? (I tried `expandable='true'` to no avail)
2. The formatting of the cards is rather ugly, lacking both margins of the text to the border, a drop shadow to the background, and the horizontal line under the title (while I can add this manually, the spacing to the title text is odd so that it is not vertically centered).

**Comment by mcanouil** [comment]
Could you share a small **self-contained "working" (reproducible)** example to work with, _i.e._, a complete Quarto document or a Git repository? Thanks.

This being said you can add any arbitrary attributes `style="height:200px"` to a callout or any fenced div.

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by joelostblom** [reply]
Thanks for the quick reply @mcanouil ! Setting the height/min-height is an improvement but the issue is that is doesn't adapt when the page width is adjusted and the content reflows. So it will either overflow on narrow pages or leave a lot of white space at the end for wide pages.

Somewhat related issue https://github.com/quarto-dev/quarto-cli/discussions/7251

Sorry for not sharing an mre. Here is the code I use currently


```qmd
:::: {layout-ncol=2}

::: {.callout-note appearance="minimal" style="height:300px"}
## Lecture learning goals

1. Understand what dashboards are
2. Explain the advantages of using dashboards
3. Define what constitutes an effective dashboard layout
4. Explain the advantages of using interactive visualizations
5. Distinguish server side from client side interactivity

:::

::: {.callout-note appearance="minimal" style="height:300px"}
## Required activities

- [Don't rush into coding](https://engineering-shiny.org/dont-rush-into-coding.html#dont-rush-into-coding) Sections 7.1.1, 7.2.1, 7.2.2.
- [UX Matters](https://engineering-shiny.org/ux-matters.html#ux-matters) Sections 6, 6.1, 6.1.1 (until the first code chunk), 6.1.2.

*You can read these either before or after class*               

:::

::::
```

![image](https://github.com/quarto-dev/quarto-cli/assets/4560057/4d561c2d-077f-49e3-a11e-0e503df953a0)

Which overflows on narrow pages:

![image](https://github.com/quarto-dev/quarto-cli/assets/4560057/4377d36e-70a6-494f-8288-f9c2943212e6)


---

`quarto check`

```
Quarto 1.4.550

[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.58.3: OK
      Deno version 1.37.2: OK

[✓] Checking versions of quarto dependencies......OK

[✓] Checking Quarto installation......OK
      Version: 1.4.550
      Path: /home/joel/miniconda3/envs/quarto/bin


[✓] Checking tools....................OK
      TinyTeX: (external install)
      Chromium: (not installed)


[✓] Checking LaTeX....................OK
      Using: TinyTex
      Path: /home/joel/.TinyTeX/bin/x86_64-linux
      Version: undefined


[✓] Checking basic markdown render....OK


[✓] Checking Python 3 installation....OK
      Version: 3.12.2 (Conda)
      Path: /home/joel/miniconda3/envs/quarto/bin/python
      Jupyter: 5.7.2
      Kernels: python3, conda-env-jb-py, (and many more, manually truncated)



[✓] Checking Jupyter engine render....OK


[✓] Checking R installation...........(None)

      Unable to locate an installed version of R.
      Install R from [1mhttps://cloud.r-project.org/[22m
```

**Reply by mcanouil** [reply]
Note that the issue is unrelated to Quarto and is pure CSS/HTML issue, which can probably be fixed by defining some rules.

**Reply by joelostblom** [reply]
I think it is related to quarto in the sense of which default rules are defined. As in the link above to Jupyter Book's grid element, that's also done through CSS but the default config is such that it works very well for creating a grid layout whereas `layout-col` in quarto does not currently. Do you have an idea of what the CSS would look like?

**Reply by mcanouil** [reply]
Use your favourite browser inspect/developer mode to see the CSS.
Note that Quarto uses Bootstrap and is "all" described in the documentation, see <https://quarto.org/docs/output-formats/html-themes.html>.
Also callouts and layout are not cards or grids, if you want to make a grid, then use a grid, see <https://quarto.org/docs/output-formats/page-layout.html#css-grid>.

**Reply by joelostblom** [reply, marked as answer]
Thanks for mentioning `grid`, this is what I was asking about in my question and seems much closer to how it is done in JupyterLab. Note that it still has the issue with the content not stretching to fill the grid. I'm not a web developer and it usually takes quite some time to figure out how to do HTML/CSS customization which is why I use Quarto/JupyterBook instead of web frameworks directly. After some fiddling it seems like all that is needed is to set `display:flex` on the grid columns; I think this would be a convenient default for quarto to set on all grids if possible similar to how it is done in JupyterBook. I'm including the setup that worked for me below if someone else comes across this looking for the same thing.

```css
.g-col-6 {
    display: flex
}
```


```qmd
::::: {.grid}

:::: {.g-col-6}

::: {.callout-note appearance="minimal"}
## Lecture learning goals

1. Motivate the need to use a dashboard framework
2. Understand the pros and cons with Dash and Shiny
3. Identify how a Dash/Shiny app is designed conceptually
4. Run a Dash/Shiny app on a local server
5. Create input widgets (dropdowns, sliders, etc)
6. Set up callbacks that update the app based on user input
7. Create an Altair/ggplot charts in dashboards
8. Update charts via input widgets

:::
    
::::

:::: {.g-col-6}

::: {.callout-note appearance="minimal"}
## Required activities

- There are several optional readings linked throughout these notes where you can learn more about Dash and Shiny.

*You can read these either before or after class*               

:::

::::

:::::
```


![image](https://github.com/quarto-dev/quarto-cli/assets/4560057/48f9132f-462b-4e67-8be4-b0f0e02c62f2)

**Reply by mcanouil** [reply]
>  I'm not a web developer and it usually takes quite some time to figure out how to do HTML/CSS customization which is why I use Quarto/JupyterBook instead of web frameworks directly.

Never Quarto or JupyterLab will ever come close to raw HTML/CSS. When you want advanced customisation, I am afraid there is no shortcut beside learning the language.
The same will apply to LaTeX, Typst, OpenXML (Word), etc.

It's not possible to map everything from other languages as the combination are nearly infinite.

>  I think this would be a convenient default 

I strongly disagree with that. It only serves your purpose while `grid` is a general framework.

**Reply by joelostblom** [reply]
> It only serves your purpose while grid is a general framework.

To be clear, I don't mean to replace the grid framework with something that only works for me, that would indeed be undesirable.

I think there might be something we are missing here, because when I read about `grid` the illustrations that are used tend to show that all items in the same row are of the same height. When adding rows/text to one of them, the default behavior seems to be that they all resize their height equally to remain aligned (as the name "grid" suggest), e.g. this link https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_grid_layout/Basic_concepts_of_grid_layout.

**Reply by mcanouil** [reply]
Your link is unrelated.
As I said, it's Bootstrap: see https://getbootstrap.com/docs/5.3/layout/css-grid/#how-it-works
Note that the link I shared also link back to Bootstrap.

Again, your use case is not general (using 1.5 for the lipsum shortcode).
The issue is not the grid.

<table>
<tr><th>Quarto document</th><th>HTML</th></tr>
<tr><td>

````qmd
---
title: "Quarto Playground"
format: html
---

This is a playground for Quarto.


::: {.grid style="border: 1px solid #ccc;"}

::: {.g-col-4 style="border: 1px solid #22b222;"}
{{< lipsum 1 >}}
:::

::: {.g-col-8 style="border: 1px solid #b22222;"}
{{< lipsum 1 >}}
:::

:::

````

</td><td>

<img width="670" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/c560d9f6-0661-4613-b008-6f0d8ece2de8">

</td></tr>
</table>

**Reply by joelostblom** [reply]
Thanks for the example, I think I get it now. What is shown as being of equal height in both the link you shared and the one I shared is the grid cell itself; not the content within the grid cell. The fact that the content I'm working with (the card/callout) has a border confused me, but I can see in this example that the border of e.g. a card is by default resized to its content rather than the available vertical space within the grid cell: https://stackblitz.com/edit/zvqkwg?file=index.html

![image](https://github.com/quarto-dev/quarto-cli/assets/4560057/03c59dc1-a651-4de7-8110-9627d0d30f47)

I could maybe get what  I am looking for by adding a border and gap to the grid itself, but I'm happy with the solution above.
